### PR TITLE
Optimization in CRUD tier

### LINKED ISSUE #1270: Optimization of Tier CRUD API and validations
**Issue/Feature Description:**

This isssues is for CRUD API pr. Some more optimization and validations can be added

**Why this issue to fixed / feature is needed(give scenarios or use cases):**

1. Make  some functions in utils to validate duplicate backends. Avoid same code in multiple places.
2. why not there is an API for getTierID()? This can be used later in some other part of code too. And once you have the ID, GET Tier API can be called #

**How to reproduce, in case of a bug:**

**Other Notes /  Environment Information: (Please give the env information, log link or any useful information for this issue)**

### PR DESCRIPTION
**Pull request by vineela1999**
**What type of PR is this?**
kind cleanup
kind enhancement

**What this PR does / why we need it**:
This PR is an optimization to CRUD tier 
when backend of a tier to be deleted  whether backend has no buckets is  checked
when tier is to be deleted the buckets of  backends of the tier should be empty check is added
**Which issue(s) this PR fixes**:

Fixes #1270 

**Test Report Added?**:
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
> /kind TESTED
> /kind NOT-TESTED

**Test Report**:

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by anvithks** [comment]
@vineela1999 Please list out the optimizations in this PR and why they are needed. 
Please add appropriate description and API output or logs where available.

**Review by anvithks — CHANGES_REQUESTED** [review]
When deleting a tier if there are backends that have buckets within then there should be an error. 
If there are no buckets then the backends can be removed and then the tier can be deleted. This can probably be handled at the Multicloud side.

**Comment by rajat-soda** [comment]
@AmitRoushan , @kumarashit @asifdxtreme : Please review.

File: api/pkg/backend/service.go [hunk 1]
```diff
@@ -664,7 +663,35 @@ func (s *APIService) UpdateTier(request *restful.Request, response *restful.Resp
...
+			return
+		}
+
+		for _, v := range res2.Buckets {
+			backendname := res1.Backend.Name
+			if backendname == v.DefaultLocation {
+				count++
+			}
+		}
+		if err != nil {
```
Comment (AmitRoushan) [inline api/pkg/backend/service.go:684]: IMO, The error check is part of ```ListBuckets```.

**Review by AmitRoushan — CHANGES_REQUESTED** [review]

File: api/pkg/backend/service.go (hunk 1 above)
Comment (vineela1999) [reply api/pkg/backend/service.go]: changed

File: api/pkg/backend/service.go [hunk 2]
```diff
@@ -346,24 +347,12 @@ func (s *APIService) DeleteBackend(request *restful.Request, response *restful.R
...
-			log.Errorf("failed to get backend details: %v\n", err)
-			response.WriteError(http.StatusInternalServerError, err)
-			return
-		}
-		backendname := res.Backend.Name
-		if backendname == v.DefaultLocation {
-			count++
-		}
-	}
+	count, err := s.HasBuckets(id, ctx)
```
Comment (rajat-soda) [inline api/pkg/backend/service.go:350]: rename to GetBucketCount

File: api/pkg/backend/service.go [hunk 3]
```diff
@@ -502,14 +491,7 @@ func (s *APIService) CreateTier(request *restful.Request, response *restful.Resp
...
 	//duplicate backends check:
-	var duplicates []string
-	visited := make(map[string]int, 0)
-	for _, val := range tier.Backends {
-		if visited[val] >= 1 {
-			duplicates = append(duplicates, val)
-		}
-		visited[val]++
-	}
+	duplicates := utils.DuplicatesCheck(tier.Backends)
```
Comment (rajat-soda) [inline api/pkg/backend/service.go:494]: rename to GetDuplicateEntities

File: api/pkg/backend/service.go [hunk 4]
```diff
@@ -589,112 +556,74 @@ func (s *APIService) UpdateTier(request *restful.Request, response *restful.Resp
...
 	}
 
-	//validation of add backends with list of backends
-	listBackendRequest := &backend.ListBackendRequest{}
-	result, err := s.backendClient.ListBackend(ctx, listBackendRequest)
-	if err != nil {
-		log.Errorf("failed to list backends: %v\n", err)
-		response.WriteError(http.StatusInternalServerError, err)
+	//check whether addBackends does not exists in the tier already
+	failAddBackends, _ := utils.ContainAndNotContain(updateTier.AddBackends, res.Tier.Backends)
```
Comment (rajat-soda) [inline api/pkg/backend/service.go:573]: replace with meaningful function name

File: api/pkg/utils/utils.go [hunk 5]
```diff
@@ -215,3 +215,32 @@ func GetTimeoutSec(objSize int64) int64 {
...
+		if visited[val] >= 1 {
+			duplicates = append(duplicates, val)
+		}
+		visited[val]++
+	}
+	return duplicates
+}
+
+//return the present values and absent values of object1 in object2
+func ContainAndNotContain(obj1 []string, obj2 []string) ([]string, []string) {
```
Comment (rajat-soda) [inline api/pkg/utils/utils.go:233]: partionObjectsBy... would be a better name

File: api/pkg/utils/utils.go [hunk 6]
```diff
@@ -215,3 +215,32 @@ func GetTimeoutSec(objSize int64) int64 {
 
 	return tmoutSec
 }
+
+//return the duplicate values present in the string array
+func DuplicatesCheck(obj []string) []string {
```
Comment (rajat-soda) [inline api/pkg/utils/utils.go:220]: Use more descriptive and meaningful variable /function names.

**Review by rajat-soda — CHANGES_REQUESTED** [review]

File: api/pkg/backend/service.go (hunk 2 above)
Comment (vineela1999) [reply api/pkg/backend/service.go]: changed

File: api/pkg/backend/service.go (hunk 3 above)
Comment (vineela1999) [reply api/pkg/backend/service.go]: changed

File: api/pkg/backend/service.go (hunk 4 above)
Comment (vineela1999) [reply api/pkg/backend/service.go]: changed

File: api/pkg/utils/utils.go (hunk 6 above)
Comment (vineela1999) [reply api/pkg/utils/utils.go]: changed

File: api/pkg/utils/utils.go (hunk 5 above)
Comment (vineela1999) [reply api/pkg/utils/utils.go]: changed

**Review by rajat-soda — APPROVED** [review]
LGTM

**Comment by rajat-soda** [comment]
Note to @kumarashit  , @NajmudheenCT   :  Merge After #1273

File: api/pkg/backend/service.go [hunk 7]
```diff
@@ -770,27 +699,85 @@ func (s *APIService) DeleteTier(request *restful.Request, response *restful.Resp
...
+
+//checking whether backend has buckets
+func (s *APIService) GetBucketsCount(backendId string, ctx context.Context) (int, error) {
+	res2, err := s.s3Client.ListBuckets(ctx, &s3.BaseRequest{})
+	if err != nil {
+		return -1, err
+	}
+
+	count := 0
+	res1, err1 := s.backendClient.GetBackend(ctx, &backend.GetBackendRequest{Id: backendId})
```
Comment (kumarashit) [inline api/pkg/backend/service.go:745]: nit: Suggestion: Give variable name meaningful than res1, res2

File: api/pkg/backend/service.go [hunk 8]
```diff
@@ -770,27 +699,85 @@ func (s *APIService) DeleteTier(request *restful.Request, response *restful.Resp
...
 	}
 	log.Info("Delete tier  successfully.")
 	response.WriteHeader(http.StatusOK)
 	return
 }
+
+//checking whether backend has buckets
+func (s *APIService) GetBucketsCount(backendId string, ctx context.Context) (int, error) {
+	res2, err := s.s3Client.ListBuckets(ctx, &s3.BaseRequest{})
+	if err != nil {
```
Comment (kumarashit) [inline api/pkg/backend/service.go:740]: log

File: api/pkg/backend/service.go [hunk 9]
```diff
@@ -770,27 +699,85 @@ func (s *APIService) DeleteTier(request *restful.Request, response *restful.Resp
...
+
+	count := 0
+	res1, err1 := s.backendClient.GetBackend(ctx, &backend.GetBackendRequest{Id: backendId})
+	if err1 != nil {
+		log.Errorf("failed to get backend details: %v\n", err)
+		return -1, err1
+	}
+
+	for _, v := range res2.Buckets {
+		backendname := res1.Backend.Name
```
Comment (kumarashit) [inline api/pkg/backend/service.go:752]: what is there is issue getting backendname or is empty result?

File: api/pkg/backend/service.go [hunk 10]
```diff
@@ -770,27 +699,85 @@ func (s *APIService) DeleteTier(request *restful.Request, response *restful.Resp
...
+		if backendname == v.DefaultLocation {
+			count++
+		}
+	}
+	return count, err
+}
+
+//give the list of invalid backends
+func (s *APIService) GetInvalidBackends(objects []string, ctx context.Context) ([]string, error) {
```
Comment (kumarashit) [inline api/pkg/backend/service.go:761]: https://github.com/sodafoundation/multi-cloud/pull/1273 has IsInvalidBackedns() function too..check and keep one

File: api/pkg/utils/utils.go [hunk 11]
```diff
@@ -215,3 +215,32 @@ func GetTimeoutSec(objSize int64) int64 {
 
 	return tmoutSec
 }
+
+//return the duplicate values present in the string array
+func GetDuplicates(list []string) []string {
```
Comment (kumarashit) [inline api/pkg/utils/utils.go:220]: check this PR https://github.com/sodafoundation/multi-cloud/pull/1273 too

**Comment by vineela1999** [comment]
pr#1273 is opened for the issue

**Pull request closed without merging**